### PR TITLE
make server_urls dynamically reloadable

### PR DESCRIPTION
**Pull request by kananindzya**
closes elastic/apm-agent-java#723 
### Checklist

- [x] Implement code
- [x] Add tests
- [ ] Update documentation
- [x] Update [CHANGELOG.md](CHANGELOG.md)
- [x] Update [supported-technologies.asciidoc](docs/supported-technologies.asciidoc)

### CODE REVIEW

**Comment by kananindzya** [comment]
> Unfortunately, it's not as easy as that. The `ApmServerClient#serverUrls` variable has to be updated as well, in a thread-safe manner. Meaning that the variable should be declared volatile.
> 
> See also [#725 (comment)](https://github.com/elastic/apm-agent-java/pull/725#issuecomment-511819759) for more details on how to use change listeners on how to register a config change callback. You can do that in the constructor of `ApmServerClient`. Also, reset the error count when the config changes.



> Also, reset the err

I added listener, but not in ApmServerClient constructor. And added
```
    public static synchronized void overrideApmServerUrls(List<URL> serverUrls) {
        logger.debug("server_urls override with value = ({}).", serverUrls);
        ApmServerClient.serverUrls = serverUrls;
    }
```
but I don't like this decision.

**Comment by kananindzya** [comment]
```
2019-07-19 16:34:43.076 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Receiving TRANSACTION event (sequence 1)
2019-07-19 16:34:43.077 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Starting new request to http://localhost:8201/intake/v2/events
2019-07-19 16:34:43.077 [apm-reporter] ERROR co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Failed to handle event of type TRANSACTION with this error: Connection refused (Connection refused)
2019-07-19 16:34:43.077 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Event handling failure
java.net.ConnectException: Connection refused (Connection refused)
        at java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
        at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
        at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
        at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.net.Socket.connect(Socket.java:589)
        at sun.net.NetworkClient.doConnect(NetworkClient.java:175)
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:463)
        at sun.net.www.http.HttpClient.openServer(HttpClient.java:558)
        at sun.net.www.http.HttpClient.<init>(HttpClient.java:242)
        at sun.net.www.http.HttpClient.New(HttpClient.java:339)
        at sun.net.www.http.HttpClient.New(HttpClient.java:357)
        at sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:1220)
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1156)
        at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1050)
        at sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:984)
        at co.elastic.apm.agent.report.IntakeV2ReportingEventHandler.startRequest(IntakeV2ReportingEventHandler.java:200)
        at co.elastic.apm.agent.report.IntakeV2ReportingEventHandler.handleEvent(IntakeV2ReportingEventHandler.java:138)
        at co.elastic.apm.agent.report.IntakeV2ReportingEventHandler.onEvent(IntakeV2ReportingEventHandler.java:116)
        at co.elastic.apm.agent.report.IntakeV2ReportingEventHandler.onEvent(IntakeV2ReportingEventHandler.java:50)
        at co.elastic.apm.agent.shaded.lmax.disruptor.BatchEventProcessor.processEvents(BatchEventProcessor.java:168)
        at co.elastic.apm.agent.shaded.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:125)
        at java.lang.Thread.run(Thread.java:748)
2019-07-19 16:34:43.078 [apm-reporter] INFO co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Backing off for 1 seconds (+/-10%)
2019-07-19 16:34:50.673 [configuration-reloader] DEBUG co.elastic.apm.agent.report.ApmServerClient - server_urls override with value = [[http://localhost:8200]].
2019-07-19 16:34:50.747 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Receiving METRICS event (sequence 2)
2019-07-19 16:34:50.748 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Starting new request to http://localhost:8200/intake/v2/events
2019-07-19 16:34:50.758 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Scheduling request timeout in 10s
2019-07-19 16:35:00.758 [apm-request-timeout-timer] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Request flush because the request timeout occurred
2019-07-19 16:35:00.759 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Receiving FLUSH event (sequence 3)
2019-07-19 16:35:00.761 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Flushing 1381 uncompressed 596 compressed bytes
^C2019-07-19 16:35:17.759 [apm-reporter] DEBUG co.elastic.apm.agent.report.IntakeV2ReportingEventHandler - Receiving SHUTDOWN event (sequence 4)
```

File: apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java [hunk 1]
```diff
@@ -73,6 +74,16 @@ public ApmServerClient(ReporterConfiguration reporterConfiguration) {
 
     public ApmServerClient(ReporterConfiguration reporterConfiguration, List<URL> serverUrls) {
         this.reporterConfiguration = reporterConfiguration;
+        for (ConfigurationOption configurationOption : reporterConfiguration.getConfigurationOptions()) {
```
Comment (felixbarny) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java]: Better create a method `public ConfigurationOption<List<URL>> getServerUrlsOption()` in `ReporterConfiguration` and then just call `reporterConfiguration.getServerUrlsOption().addChangeListener(...)` here

File: apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java [hunk 2]
```diff
@@ -245,4 +256,10 @@ int getErrorCount() {
         T withConnection(HttpURLConnection connection) throws IOException;
     }
 
+    public synchronized void overrideApmServerUrls(List<URL> serverUrls) {
```
Comment (felixbarny) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java]: Just inline this to the `onChange` method. Doesn't have to be synchronized.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java [hunk 3]
```diff
@@ -245,4 +256,10 @@ int getErrorCount() {
...
+        logger.debug("server_urls override with value = ({}).", serverUrls);
+        this.serverUrls = serverUrls;
```
Comment (felixbarny) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java]: Shuffle the URLs by reusing the `shuffleUrls` method. To achieve that, change the parameter of `shuffleUrls` from `ReporterConfiguration` to `List<URL>`.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java (hunk 1 above)
Comment (kananindzya) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java]: fixed

File: apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java (hunk 2 above)
Comment (kananindzya) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java]: fixed

File: apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java (hunk 3 above)
Comment (kananindzya) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java]: added

**Comment by kananindzya** [comment]
@felixbarny  I fixed according to comments

**Comment by felixbarny** [comment]
could you add a section to the changelog?

**Comment by kananindzya** [comment]
> could you add a section to the changelog?

added

**Comment by felixbarny** [comment]
Easier than using an actual properties file for testing is to use an in-memory PropertySource like here:
https://github.com/elastic/apm-agent-java/blob/45857cea69a82ac97d5f1bee266f58fbedddb43e/apm-agent-plugins/apm-slf4j-plugin/src/test/java/co/elastic/apm/agent/slf4j/Slf4JMdcActivationListenerTest.java#L175

File: apm-agent-core/src/test/java/co/elastic/apm/agent/report/ApmServerClientTest.java [hunk 4]
```diff
@@ -24,18 +24,30 @@
...
 
 import java.io.FileNotFoundException;
+import java.io.IOException;
 import java.net.HttpURLConnection;
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.net.URLConnection;
+import java.util.Arrays;
 import java.util.List;
+import java.util.ServiceLoader;
```
Comment (eyalkoren) [inline apm-agent-core/src/test/java/co/elastic/apm/agent/report/ApmServerClientTest.java]: remove unused imports

File: apm-agent-core/src/test/java/co/elastic/apm/agent/report/ApmServerClientTest.java [hunk 5]
```diff
@@ -52,17 +64,27 @@
     @Rule
     public WireMockRule apmServer2 = new WireMockRule(WireMockConfiguration.wireMockConfig().dynamicPort());
     private ApmServerClient apmServerClient;
+    protected static ConfigurationRegistry config;
```
Comment (eyalkoren) [inline apm-agent-core/src/test/java/co/elastic/apm/agent/report/ApmServerClientTest.java]:
```suggestion
    private ConfigurationRegistry config;
```

File: apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java [hunk 6]
```diff
@@ -240,6 +251,10 @@ int getErrorCount() {
         return errorCount.get();
     }
 
+    protected List<URL> getServerUrls()  {
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerClient.java]:
```suggestion
   List<URL> getServerUrls()  {
```

**Comment by kananindzya** [comment]
> `getServerUrls`

hi @eyalkoren , I added protected ApmServerClient#getServerUrls, and added test with using
```
config.save("server_urls", tempUrl.toString(), SpyConfiguration.CONFIG_SOURCE_NAME);
```

File: apm-agent-core/src/test/java/co/elastic/apm/agent/report/ApmServerClientTest.java (hunk 4 above)
Comment (kananindzya) [reply apm-agent-core/src/test/java/co/elastic/apm/agent/report/ApmServerClientTest.java]: removed

**Comment by felixbarny** [comment]
can you merge in the changes from master again? that should fix the test failures

**Comment by felixbarny** [comment]
run the tests

**Pull request merged**